ext/ddtrace_profiling_native_extension/collectors_stack.c

```diff
@@ -171,6 +178,8 @@ void sample_thread(
   rb_raise(rb_eArgError, "Unexpected value for sample_type: %d", type);
 }
 
+#define CHARSLICE_EQUALS(must_be_a_literal, charslice) (strlen("" must_be_a_literal) == charslice.len && strncmp(must_be_a_literal, charslice.ptr, charslice.len) == 0)
```

Comment:
So this is a bit of C dark magic cool trick. It relies on the fact that you can do this in C:

```c
char *foo = "hello " "world";
```

And the result is concatenated. Note that you can only do this for literals, that is, it's not a generic way of joining strings in C.

In this case, I wanted to force the string being compared to the charslice to be a literal, so by by doing 

```
strlen("" must_be_a_literal)
```

This causes the compiler to barf if you do this:
```c
        char *sleep_string = "sleep";
        if (CHARSLICE_EQUALS(sleep_string, name_slice)) { 
```

Output:

```
../../../../ext/ddtrace_profiling_native_extension/collectors_stack.c: In function ‘sample_thread_internal’:
../../../../ext/ddtrace_profiling_native_extension/collectors_stack.c:288:30: error: expected ‘)’ before ‘sleep_string’
  288 |         if (CHARSLICE_EQUALS(sleep_string, name_slice)) {
      |                              ^~~~~~~~~~~~
../../../../ext/ddtrace_profiling_native_extension/collectors_stack.c:181:67: note: in definition of macro ‘CHARSLICE_EQUALS’
  181 | #define CHARSLICE_EQUALS(must_be_a_literal, charslice) (strlen("" must_be_a_literal) == charslice.len && strncmp(must_be_a_literal, charslice.ptr, charslice.len) == 0)
      |                                                                   ^~~~~~~~~~~~~~~~~
../../../../ext/ddtrace_profiling_native_extension/collectors_stack.c:181:63: note: to match this ‘(’
  181 | #define CHARSLICE_EQUALS(must_be_a_literal, charslice) (strlen("" must_be_a_literal) == charslice.len && strncmp(must_be_a_literal, charslice.ptr, charslice.len) == 0)
      |                                                               ^
../../../../ext/ddtrace_profiling_native_extension/collectors_stack.c:288:13: note: in expansion of macro ‘CHARSLICE_EQUALS’
  288 |         if (CHARSLICE_EQUALS(sleep_string, name_slice)) {
      |             ^~~~~~~~~~~~~~~~
```

The error could be a bit better but I hope the "must_be_a_literal" hint would lead people in the right direction ;)

> I'd expect to need one in both places or neither, what's the deal here? 😄

The concat doesn't need to happen on both because `"" must_be_a_literal` doesn't do anything else other than cause the error to happen, and otherwise is a no-op. (Doing it on both causes more errors on the compiler side, so from that POV is even slightly worse)

---

As a final note, why I did this: more out of a tiny cheap micro-optimization; by forcing it to be a literal, I know the compiler can optimize away the `strlen`; otherwise I would probably have wanted to cache it somewhere, which would be more boilerplate.